src/protocol/ChatChannel.cpp

```diff
@@ -0,0 +1,196 @@
+        return false;
+    }
+
+    if (connection()->findChannel<ChatChannel>(Channel::Inbound)) {
+        qDebug() << "Rejecting request for" << type() << "channel because one is already open";
+        return false;
+    }
+
+    return true;
+}
+
+bool ChatChannel::allowOutboundChannelRequest(Data::Control::OpenChannel *request)
+{
+    Q_UNUSED(request);
+
+    if (connection()->findChannel<ChatChannel>(Channel::Outbound)) {
+        BUG() << "Rejecting outbound request for" << type() << "channel because one is already open on this connection";
+        return false;
+    }
+
+    if (connection()->purpose() != Connection::Purpose::KnownContact) {
+        BUG() << "Rejecting outbound request for" << type() << "channel for connection with unexpected purpose" << int(connection()->purpose());
+        return false;
+    }
+
+    return true;
+}
+
+void ChatChannel::receivePacket(const QByteArray &packet)
+{
+    Data::Chat::Packet message;
+    if (!message.ParseFromArray(packet.constData(), packet.size())) {
+        closeChannel();
+        return;
+    }
+
+    if (message.has_chat_message()) {
+        handleChatMessage(message.chat_message());
+    } else if (message.has_chat_acknowledge()) {
+        handleChatAcknowledge(message.chat_acknowledge());
+    } else {
+        qWarning() << "Unrecognized message on" << type();
+        closeChannel();
+    }
+}
+
+bool ChatChannel::sendChatMessage(QString text, QDateTime time, MessageId &id)
+{
+    id = ++lastMessageId;
+    return sendChatMessageWithId(text, time, id);
+}
+
+bool ChatChannel::sendChatMessageWithId(QString text, QDateTime time, MessageId id)
+{
+    if (direction() != Outbound) {
+        BUG() << "Chat channels are unidirectional, and this is not an outbound channel";
+        return false;
+    }
+
+    QScopedPointer<Data::Chat::ChatMessage> message(new Data::Chat::ChatMessage);
+    message->set_message_id(id);
+
+    if (text.isEmpty()) {
+        BUG() << "Chat message is empty, and it should've been discarded";
+        return false;
+    } else if (text.size() > MessageMaxCharacters) {
+        BUG() << "Chat message is too long (" << text.size() << "characters), and it should've been limited already. Truncated.";
+        text.truncate(MessageMaxCharacters);
+    }
+
+    // Also converts to UTF-8
+    message->set_message_text(text.toStdString());
+
+    if (!time.isNull())
+        message->set_time_delta(qMin(QDateTime::currentDateTime().secsTo(time), qint64(0)));
+
+    Data::Chat::Packet packet;
+    packet.set_allocated_chat_message(message.take());
+    if (!Channel::sendMessage(packet))
+        return false;
+
+    pendingMessages.insert(id);
+    return true;
+}
+
+void ChatChannel::handleChatMessage(const Data::Chat::ChatMessage &message)
+{
+    QScopedPointer<Data::Chat::ChatAcknowledge> response(new Data::Chat::ChatAcknowledge);
+
+    // QString::fromStdString decodes the string as UTF-8, replacing all invalid sequences and
```

Comment:
This is documented as using fromAscii. I know that the functionality *nowdays* will use UTF8 conversion, but it might be smarter to be explicit about this, using fromUtf8